### PR TITLE
 Merge bitcoin#13134: net: Add option `-enablebip61` to configure sending of BIP61 notifications

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
Tests currently fail. I did my best to fix them, this is a dependency for further p2p refactoring and changes I have done.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -3502,10 +3510,12 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman)
     AssertLockHeld(cs_main);
     CNodeState &state = *State(pnode->GetId());
 
-    for (const CBlockReject& reject : state.rejects) {
-        // Pass allowOptimisticSend=true here to allow sending the reject message immediately instead of waiting for the
+    if (g_enable_bip61) {
+        for (const CBlockReject& reject : state.rejects) {
+            // Pass allowOptimisticSend=true here to allow sending the reject message immediately instead of waiting for the
         // network thread to do it. This gives us a chance of actually sending out the reject before we close the connection.
         connman->PushMessage(pnode, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::REJECT, (std::string)NetMsgType::BLOCK, reject.chRejectCode, reject.strRejectReason, reject.hashBlock), true);
```
Comment (UdjinM6) [inline src/net_processing.cpp:3517]: should fix indents here

File: src/net_processing.h [hunk 2]
```diff
@@ -37,6 +37,11 @@ static constexpr int64_t EXTRA_PEER_CHECK_INTERVAL = 45;
 /** Minimum time an outbound-peer-eviction candidate must be connected for, in order to evict, in seconds */
 static constexpr int64_t MINIMUM_CONNECT_TIME = 30;
 
+/** Default for BIP61 (sending reject messages) */
+static constexpr bool DEFAULT_ENABLE_BIP61 = true;
```
Comment (thephez) [inline src/net_processing.h:41]: General comment regarding future BIP61-related backports. IIRC, Bitcoin eventually sets this to false and then entirely removes BIP61 support. The removal of BIP61 from Bitcoin Core was extremely controversial from the standpoint of mobile wallets. 

Not sure what our plan is for when we get to that point, but wanted to bring it up so we do not blindly follow Bitcoin. I imagine @hashengineering and @quantumexplorer will want to provide input on that.

Comment (PastaPastaPasta) [reply src/net_processing.h:41]: Correct, I've seen this, but these backports at this point are needed for resolving conflicts. Also, I do think their reasoning for removing it made senes

File: src/net_processing.cpp (hunk 1 above)
Comment (PastaPastaPasta) [reply src/net_processing.cpp]: https://github.com/dashpay/dash/pull/3414/commits/84077095558f72d3ebaeb29f070c6fa8e5b3f30f

File: test/functional/invalidtxrequest.py [hunk 3]
```diff
@@ -81,7 +84,7 @@ def run_test(self):
         # and we get disconnected immediately
         self.log.info('Test a transaction that is rejected')
         tx1 = create_transaction(block1.vtx[0], 0, b'\x64', 50 * COIN)
-        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True)
+        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True, reject_code=REJECT_INVALID, reject_reason=b'mandatory-script-verify-flag-failed (Invalid OP_IF construction)')
```
[the remark this answers is not in the document]
Comment (codablock) [reply test/functional/invalidtxrequest.py:87]: Yeah I've seen this behavior as well with other stuff in the past. Can't look into it right now, but I remember it had something to do with Dash not doing "optimistic send" anymore and thus closes the connection before the socket handler thread sends the reject message. Maybe the solution is to wait for disconnect a few milliseconds when the send buffer is non-empty to give it a chance to send out a few reject messages first.

[the remark this answers is not in the document]
Comment (codablock) [reply test/functional/invalidtxrequest.py:87]: I can implement this fix later today if I find time.

[the remark this answers is not in the document]
Comment (codablock) [reply test/functional/invalidtxrequest.py:87]: @UdjinM6 see https://github.com/dashpay/dash/pull/3429

File: src/net_processing.h (hunk 2 above)
Comment (QuantumExplorer) [reply src/net_processing.h:41]: Yeah I strongly encourage the continued use of BIP61, the reasoning to remove it was too weak in Bitcoin.

File: test/functional/invalidtxrequest.py [hunk 4]
```diff
@@ -94,6 +97,14 @@ def run_test(self):
...
         self.test_orphan_tx_handling(block2.vtx[0].sha256, True)
 
+        # restart node with sending BIP61 messages disabled, check that it disconnects without sending the reject message
+        self.log.info('Test a transaction that is rejected, with BIP61 disabled')
+        self.restart_node(0, ['-enablebip61=0','-persistmempool=0'])
+        self.reconnect_p2p(num_connections=1)
+        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True)
+        # send_txs_and_test will have waited for disconnect, so we can safely check that no reject has been received
+        assert_equal(node.p2p.reject_code_received, None)
+
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply test/functional/invalidtxrequest.py:104]: https://github.com/dashpay/dash/pull/3414/commits/a46483ed47dbf1e1d780ab1b1c87286859e142e3

**Comment by UdjinM6** [comment]
Needs rebase

**Comment by PastaPastaPasta** [comment]
rebased on develop

**Comment by PastaPastaPasta** [comment]
rebased again to (hopefully) resolve tests

**Pull request merged**